app/src/main/java/org/mozilla/reference/browser/NotificationManager.kt

```diff
@@ -0,0 +1,126 @@
+
+/**
+ * Manages notification channels and allows displaying different types of notifications.
+ */
+class NotificationManager(private val context: Context) {
+    companion object {
+        const val RECEIVE_TABS_TAG = "ReceivedTabs"
+        const val RECEIVE_TABS_CHANNEL_ID = "ReceivedTabsChannel"
+    }
+
+    init {
+        // Create the notification channels we are going to use, but only on API 26+ because the NotificationChannel
+        // class is new and not in the support library.
+        if (SDK_INT >= Build.VERSION_CODES.O) {
+            createNotificationChannel(
+                RECEIVE_TABS_CHANNEL_ID,
+                // Pick 'high' because this is a user-triggered action that is expected to be part of a continuity flow.
+                // That is, user is expected to be waiting for this notification on their device; make it obvious.
+                AndroidNotificationManager.IMPORTANCE_HIGH,
+                // Name and description are shown in the 'app notifications' settings for the app.
+                context.getString(R.string.fxa_received_tab_channel_name),
+                context.getString(R.string.fxa_received_tab_channel_description)
+            )
+        }
+    }
+
+    private val logger = Logger("NotificationManager")
+    fun showReceivedTabs(device: Device?, tabs: List<TabData>) {
+        // In the future, experiment with displaying multiple tabs from the same device as as Notification Groups.
+        // For now, a single notification per tab received will suffice.
+        logger.debug("Showing ${tabs.size} tab(s) received from deviceID=${device?.id}")
+        tabs.forEach { tab ->
+            val intent = Intent(Intent.ACTION_VIEW, Uri.parse(tab.url)).apply {
+                flags = Intent.FLAG_ACTIVITY_NEW_TASK
+            }
+            val pendingIntent: PendingIntent =
+                PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_ONE_SHOT)
+
+            val builder = NotificationCompat.Builder(context, RECEIVE_TABS_CHANNEL_ID)
+                .setSmallIcon(R.drawable.ic_status_logo)
+                .setSendTabTitle(context, device, tab)
+                .setWhen(System.currentTimeMillis())
+                .setContentText(tab.url)
+                .setContentIntent(pendingIntent)
+                .setAutoCancel(true)
+                .setPriority(NotificationCompat.PRIORITY_HIGH)
+                .setDefaults(Notification.DEFAULT_VIBRATE or Notification.DEFAULT_SOUND)
+
+            if (SDK_INT >= Build.VERSION_CODES.M) {
+                builder.setCategory(Notification.CATEGORY_REMINDER)
+            }
+
+            // Pick a random ID for this notification so that different tabs do not clash.
+            @SuppressWarnings("MagicNumber")
+            val notificationId = (Math.random() * 100).toInt()
+
+            with(NotificationManagerCompat.from(context)) {
+                notify(RECEIVE_TABS_TAG, notificationId, builder.build())
+            }
```

Comment:
What's the advantage of `with()` here. Can't you call notify directly on it?

```kotlin
NotificationManagerCompat.from(context)
    .notify(..)
```

Reply:
I think it was taken from this Android developer guide 🙃  : https://developer.android.com/training/notify-user/build-notification#notify